src/lib.rs

```diff
@@ -84,18 +118,55 @@ mod mem {
 
     use ::size_of;
 
+    #[cfg_attr(
+        any(test, feature = "pre"),
+        pre::pre(valid_ptr(ptr, w)),
+        pre::pre("`ptr` points to a single allocation that is valid for at least `count` bytes"),
+        pre::pre(count <= std::isize::MAX as usize)
+    )]
     #[inline(never)]
-    pub fn zero<T: Sized + Copy>(slice: &mut [T]) {
-        let ptr = slice.as_mut_ptr() as *mut u8;
-        for i in 0 .. size_of(slice) {
-            unsafe {
-                std::ptr::write_volatile(ptr.offset(i as isize), 0);
-            }
+    pub unsafe fn zero(ptr: *mut u8, count: usize) {
+        for i in 0 .. count {
+            #[cfg_attr(
+                any(test, feature = "pre"),
+                forward(impl pre::std::mut_pointer),
+                assure(
+                    "the starting and the resulting pointer are in bounds of the same allocated object",
+                    reason = "this is guaranteed by the precondition to this function"
+                ),
+                assure(
+                    "the computed offset, in bytes, does not overflow an `isize`",
+                    reason = "`computed offset <= count <= isize::MAX`"
+                ),
+                assure(
+                    "performing the offset does not result in overflow",
+                    reason = "a single allocation does not rely on overflow to index all elements and `i as isize >= 0`"
+                )
+            )]
+            let offset_ptr = ptr.offset(i as isize);
+
+            #[cfg_attr(
+                any(test, feature = "pre"),
+                forward(pre),
+                assure(
+                    valid_ptr(dst, w),
+                    reason = "the call to offset above produced a valid pointer into the allocation"
+                ),
+                assure(
+                    proper_align(dst),
+                    reason = "`align_of::<*mut u8>() == 1` and any pointer has an alignment of `1`"
+                )
+            )]
+            std::ptr::write_volatile(offset_ptr, 0);
         }
     }
 
     #[inline(never)]
-    pub fn cmp<T: Sized + Copy>(us: &[T], them: &[T]) -> bool {
+    #[cfg_attr(
+        any(test, feature = "pre"),
+        pre::pre("`T` does not have any padding bytes")
+    )]
+    pub unsafe fn cmp<T: Sized + Copy>(us: &[T], them: &[T]) -> bool {
```

Comment:
If `zero` is working on `u8` now, why not make this `u8` too? Actually this was really silly, taking any `&[T]` to use as `&[u8]` results in code bloat (monomorphizing into the exact same thing for all used `T`) :D

Reply:
The problem with that is that `&[u8]` may be UB in case `T` has padding bytes. Ideally `*const [T]` is best suited for that kind of thing, but usage of that is not really stabilized. If at least the `cmp` function should be UB-free (to have UB in less places), it should take `us: *const u8, us_len: usize, them: *const u8, then_len: usize` as arguments. I can make that change though.

Reply:
I didn't mean actually using `&[u8]`, I meant that the `T` inside is basically only used as bytes! Of course the signature should take raw pointers, just like `fn zero`.